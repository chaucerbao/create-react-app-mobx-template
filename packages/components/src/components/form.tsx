// TODO: Consider a better way to calculate `fieldValue()` for CheckboxGroup

// Dependencies
import React from 'react'

// Type definitions
interface FieldProps {
  children: React.ReactNode
  error?: string
  label?: string
  name: string
}
interface GroupProps {
  children: React.ReactNode
  error?: string
  label?: string
}
type FormFieldProps = Pick<FieldProps, Exclude<keyof FieldProps, 'children'>>
type FormGroupProps = Pick<GroupProps, Exclude<keyof GroupProps, 'children'>>
interface OptionProps {
  options: Array<{ label: string; value: string }>
}
interface FileUploadProps {
  render?: () => React.ReactNode
  renderPreview?: (
    key: string,
    Image: React.StatelessComponent<{ [key: string]: any }>,
    file: File
  ) => React.ReactNode
}
export type ChangeEvent = React.FormEvent<
  HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
>

// Components
const Field = ({ children, error, label, name }: FieldProps) => (
  <div>
    <label htmlFor={name}>
      {label && <span>{label}</span>}
      {children}
    </label>
    {error && <span>{error}</span>}
  </div>
)

const Group = ({ children, error, label }: GroupProps) => (
  <div>
    {label && <span>{label}</span>}
    {error && <span>{error}</span>}
    {children}
  </div>
)

const Input = ({
  error,
  label,
  ...props
}: FormFieldProps & React.HTMLProps<HTMLInputElement>) => (
  <Field error={error} label={label} name={props.name}>
    <input {...props} id={props.name} type="text" />
  </Field>
)

const TextArea = ({
  error,
  label,
  ...props
}: FormFieldProps & React.HTMLProps<HTMLTextAreaElement>) => (
  <Field error={error} label={label} name={props.name}>
    <textarea {...props} id={props.name} />
  </Field>
)

const Select = ({
  error,
  label,
  options,
  ...props
}: FormFieldProps & React.HTMLProps<HTMLSelectElement> & OptionProps) => (
  <Field error={error} label={label} name={props.name}>
    <select {...props} id={props.name}>
      {options.map(({ label, value }) => (
        <option key={`${props.name}:${value}`} value={value}>
          {label}
        </option>
      ))}
    </select>
  </Field>
)

const Checkbox = ({
  error,
  label,
  ...props
}: FormFieldProps & React.HTMLProps<HTMLInputElement>) => (
  <Field error={error} label={label} name={props.name}>
    <input {...props} id={props.name} type="checkbox" />
  </Field>
)

const CheckboxGroup = ({
  error,
  label,
  name,
  options,
  ...props
}: FormGroupProps & React.HTMLProps<HTMLInputElement> & OptionProps) => (
  <Group error={error} label={label}>
    {options.map(({ label, value }) => (
      <label key={`${name}:${value}`} htmlFor={`${name}:${value}`}>
        <input
          {...props}
          id={`${name}:${value}`}
          name={name}
          value={value}
          type="checkbox"
          checked={
            Array.isArray(props.value) && props.value.indexOf(value) > -1
          }
          data-checkbox-group
        />
        <span>{label}</span>
      </label>
    ))}
  </Group>
)

const RadioGroup = ({
  error,
  label,
  name,
  options,
  ...props
}: FormGroupProps & React.HTMLProps<HTMLInputElement> & OptionProps) => (
  <Group error={error} label={label}>
    {options.map(({ label, value }) => (
      <label key={`${name}:${value}`} htmlFor={`${name}:${value}`}>
        <input
          {...props}
          id={`${name}:${value}`}
          name={name}
          value={value}
          type="radio"
        />
        <span>{label}</span>
      </label>
    ))}
  </Group>
)

class FileUpload extends React.Component<
  FormFieldProps & FileUploadProps & React.HTMLProps<HTMLInputElement>
> {
  input?: HTMLInputElement
  imageFiles: File[] = []
  imagePreviews: HTMLImageElement[] = []

  componentWillUpdate() {
    const { input } = this

    const imageFiles =
      input && input.files
        ? Array.from(input.files).filter(file =>
            /\.(gif|jpg|png)$/.test(file.name)
          )
        : []

    if (imageFiles) {
      imageFiles.forEach((file, i) => {
        const fileReader = new FileReader()

        fileReader.addEventListener('load', () => {
          if (this.imagePreviews[i])
            this.imagePreviews[i].src = fileReader.result
        })
        fileReader.readAsDataURL(file)
      })
    }

    this.imageFiles = imageFiles
    this.imagePreviews = []
  }

  render() {
    const { error, label, render, renderPreview, ...props } = this.props

    return (
      <Field error={error} label={label} name={props.name}>
        {render && render()}
        {renderPreview &&
          this.imageFiles.map((file, i) =>
            renderPreview(
              `${props.name}:${i}`,
              imageProps => (
                <img
                  {...imageProps}
                  ref={img => (this.imagePreviews[i] = img!)}
                />
              ),
              file
            )
          )}
        <input
          {...props}
          id={props.name}
          type="file"
          ref={input => (this.input = input!)}
        />
      </Field>
    )
  }
}

function fieldValue(e: ChangeEvent) {
  const field = e.currentTarget

  if (field instanceof HTMLSelectElement && field.multiple) {
    return Array.from(field.options)
      .filter(option => option.selected)
      .map(option => option.value)
  } else if (field instanceof HTMLInputElement && field.type === 'checkbox') {
    if (field.dataset.checkboxGroup) {
      return Array.from(
        field.parentElement!.parentElement!.querySelectorAll('input:checked')
      ).map(input => (input as HTMLInputElement).value)
    } else {
      return field.checked
    }
  } else if (
    field instanceof HTMLInputElement &&
    field.type === 'file' &&
    field.files
  ) {
    return field.multiple ? Array.from(field.files) : field.files[0]
  }

  return field.value
}

// Exports
export {
  Input,
  TextArea,
  Select,
  Checkbox,
  CheckboxGroup,
  RadioGroup,
  FileUpload,
  fieldValue
}
