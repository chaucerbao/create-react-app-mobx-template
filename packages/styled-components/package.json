{
  "name": "styled-components-styles",
  "version": "0.1.0",
  "description": "Using styled-components to manage the styles",
  "author": "Alvin Chan <no@email.com>",
  "license": "ISC",
  "stylelint": {
    "processors": [
      "stylelint-processor-styled-components"
    ],
    "extends": [
      "stylelint-config-standard",
      "stylelint-config-concentric-order",
      "stylelint-config-styled-components"
    ]
  },
  "devDependencies": {
    "source-map-explorer": "^1.5.0",
    "stylelint": "^9.2.0",
    "stylelint-config-concentric-order": "^3.1.0",
    "stylelint-config-standard": "^18.2.0",
    "stylelint-config-styled-components": "^0.1.1",
    "stylelint-processor-styled-components": "^1.3.1",
    "tslint": "^5.10.0",
    "webpack-cli": "^2.1.3",
    "webpack-dev-server": "^3.1.4"
  },
  "dependencies": {
    "styled-components": "^3.2.6"
  },
  "scripts": {
    "start": "webpack-dev-server --config ../../webpack.config.js -d",
    "build": "webpack-cli --config ../../webpack.config.js -p --devtool source-map --progress",
    "analyze": "source-map-explorer public/js/script.js{,.map}"
  }
}
