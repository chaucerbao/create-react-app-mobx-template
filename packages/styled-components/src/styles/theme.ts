// tslint:disable object-literal-sort-keys

// Type definitions
export interface ITheme {
  breakpoints: IDictionary
  colors: IDictionary
  spacers: IDictionary
}
interface IDictionary {
  [key: string]: number | string
}

// Theme
const breakpoints = {
  mobile: '0',
  tablet: '768px',
  desktop: '1280px',
}

const colors = {
  primary: '#444',
  grey: '#888',
  white: '#fafafa',
}

const spacers = {
  xs: '4px',
  sm: '8px',
  md: '16px',
  lg: '24px',
  xl: '32px',
}

const sort = (dictionary: IDictionary) =>
  Object.entries(dictionary)
    .sort((a, b) => {
      const aValue = parseFloat(a[1].toString())
      const bValue = parseFloat(b[1].toString())

      if (aValue < bValue) {
        return -1
      }
      if (aValue > bValue) {
        return 1
      }

      return 0
    })
    .reduce((sortedDictionary: IDictionary, definition) => {
      sortedDictionary[definition[0]] = definition[1]

      return sortedDictionary
    }, {})

const theme: ITheme = {
  breakpoints: sort(breakpoints),
  colors,
  spacers: sort(spacers),
}

// Exports
export default theme
