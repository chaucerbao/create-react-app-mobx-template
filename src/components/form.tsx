// Third-party dependencies
import * as React from 'react'
import styled, { css } from 'styled-components'

// Type definitions
interface IStyledComponent {
  ref?: any
}
interface IField {
  error?: string
  label?: string
  name: string
}
interface IOptionSet {
  options: IOption[]
}
interface IOption {
  label: string
  value: string
}
type IInput = IField & IStyledComponent & React.HTMLProps<HTMLInputElement>
type ITextArea = IField &
  IStyledComponent &
  React.HTMLProps<HTMLTextAreaElement>
type ISelect = IField &
  IStyledComponent &
  React.HTMLProps<HTMLSelectElement> &
  IOptionSet
type ICheckbox = IInput & IOptionSet

// Components
const Field = ({
  children,
  error,
  label,
  name
}: IField & { children: any }) => (
  <StyledField>
    {label && <FieldLabel htmlFor={name}>{label}</FieldLabel>}
    {children}
    {error && <ErrorMessage>{error}</ErrorMessage>}
  </StyledField>
)

const Input = ({ error, label, name, type = 'text', ...props }: IInput) => (
  <Field label={label} name={name} error={error}>
    <StyledInput {...props} id={name} name={name} type={type} />
  </Field>
)

const TextArea = ({ error, label, name, ...props }: ITextArea) => (
  <Field label={label} name={name} error={error}>
    <StyledTextArea {...props} id={name} name={name} />
  </Field>
)

const Select = ({
  error,
  label,
  name,
  value,
  options = [],
  ...props
}: ISelect) => (
  <Field label={label} name={name} error={error}>
    <SelectWrapper>
      <StyledSelect {...props} id={name} name={name} value={value}>
        {options.map(option => (
          <option key={`${name}:${option.label}`} value={option.value}>
            {option.label}
          </option>
        ))}
      </StyledSelect>
    </SelectWrapper>
  </Field>
)

const Checkbox = ({
  error,
  label,
  name,
  value,
  type = 'checkbox',
  options = [],
  ...props
}: ICheckbox) => (
  <Field label={label} name={name} error={error}>
    {options.map(option => (
      <label key={`${name}:${option.label}`}>
        <CheckboxField
          {...props}
          name={name}
          value={option.value}
          type={type}
          checked={
            Array.isArray(value)
              ? value.indexOf(option.value.toString()) > -1
              : value === option.value
          }
        />
        <CheckboxLabel>{option.label}</CheckboxLabel>
      </label>
    ))}
  </Field>
)

const Radio = (props: ICheckbox) => <Checkbox {...props} type="radio" />

// Styles
const StyledField = styled.div`
  display: flex;
  flex-flow: column;
`
const FieldLabel = styled.label`
  font-weight: bold;
`
const ErrorMessage = styled.span`
  color: ${props => props.theme.red};
`
const inputBox = css`
  margin: 0;
  outline: none;
  border: 1px solid ${props => props.theme.darkgray};
  border-radius: 0;
  background: ${props => props.theme.white};
  padding: 4px;
  &:focus {
    border-color: ${props => props.theme.darkblue};
  }
`
const StyledInput = styled.input`
  ${inputBox};
`
const StyledTextArea = styled.textarea`
  ${inputBox};
`
const StyledSelect = styled.select`
  ${inputBox};
  appearance: none;
  flex: 1;
`
const SelectWrapper = styled.div`
  display: flex;
  position: relative;
  &::after {
    display: block;
    position: absolute;
    top: calc(50% - 5px / 2);
    right: 5px;
    border: 5px solid;
    border-color: ${props => props.theme.darkgray} transparent transparent
      transparent;
    pointer-events: none;
    content: '';
  }
`
const CheckboxLabel = styled.span`
  display: flex;
  position: relative;
  align-items: center;
  &::before {
    ${inputBox};
    display: block;
    margin-right: 4px;
    padding: 0;
    width: 13px;
    height: 13px;
    content: '';
  }
  &::after {
    position: absolute;
    top: calc(50% - 11px / 2);
    left: 2px;
    transform: scale(0);
    transition: transform 0.2s;
    background: ${props => props.theme.darkblue};
    width: 11px;
    height: 11px;
    content: '';
  }
`
const CheckboxField = styled.input`
  position: absolute;
  opacity: 0;
  &:checked + span::after {
    transform: scale(1);
  }
  & + span::before,
  &:checked + span::after {
    border-radius: ${props => (props.type === 'radio' ? '50%' : '0')};
  }
  &:focus + span::before {
    border-color: ${props => props.theme.darkblue};
  }
`

// Exports
export { Checkbox, Input, Radio, Select, TextArea }
