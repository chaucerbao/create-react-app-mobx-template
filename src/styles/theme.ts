// Styles
export default {
  black: '#000',
  blue: 'steelblue',
  darkblue: 'royalblue',
  darkgray: '#333',
  dimgray: 'dimgray',
  lightblue: 'lightsteelblue',
  lightgray: 'gainsboro',
  red: 'red',
  white: '#fff'
}
