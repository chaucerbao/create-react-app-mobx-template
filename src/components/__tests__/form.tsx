// Third-party dependencies
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

// Local dependencies
import theme from '../../styles/theme'

// Component
import { Checkbox, Input, Radio, Select, TextArea } from '../form'

describe('<Input />', () => {
  it('renders', () => {
    const component = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Input
            label="Label"
            name="name"
            value="Value"
            error="Error message"
          />
        </ThemeProvider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})

describe('<TextArea />', () => {
  it('renders', () => {
    const component = renderer
      .create(
        <ThemeProvider theme={theme}>
          <TextArea
            label="Label"
            name="name"
            value="Value"
            error="Error message"
          />
        </ThemeProvider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})

describe('<Select />', () => {
  it('renders', () => {
    const component = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Select
            label="Label"
            name="name"
            value="two"
            options={[
              { label: 'Option 1', value: 'one' },
              { label: 'Option 2', value: 'two' },
              { label: 'Option 3', value: 'three' }
            ]}
            error="Error message"
          />
        </ThemeProvider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})

describe('<Checkbox />', () => {
  it('renders', () => {
    const component = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Checkbox
            label="Label"
            name="name"
            value={['one', 'three']}
            options={[
              { label: 'Option 1', value: 'one' },
              { label: 'Option 2', value: 'two' },
              { label: 'Option 3', value: 'three' }
            ]}
            error="Error message"
          />
        </ThemeProvider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})

describe('<Radio />', () => {
  it('renders', () => {
    const component = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Radio
            label="Label"
            name="name"
            value="two"
            options={[
              { label: 'Option 1', value: 'one' },
              { label: 'Option 2', value: 'two' },
              { label: 'Option 3', value: 'three' }
            ]}
            error="Error message"
          />
        </ThemeProvider>
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })
})
