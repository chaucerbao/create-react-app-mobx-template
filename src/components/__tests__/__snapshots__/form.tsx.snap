// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
}

.c1 {
  font-weight: bold;
}

.c4 {
  color: red;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3::before {
  margin: 0;
  outline: none;
  border: 1px solid #333;
  border-radius: 0;
  background: #fff;
  padding: 4px;
  display: block;
  margin-right: 4px;
  padding: 0;
  width: 13px;
  height: 13px;
  content: '';
}

.c3::before:focus {
  border-color: royalblue;
}

.c3::after {
  position: absolute;
  top: calc(50% - 11px / 2);
  left: 2px;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  background: royalblue;
  width: 11px;
  height: 11px;
  content: '';
}

.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2:checked + span::after {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c2 + span::before,
.c2:checked + span::after {
  border-radius: 0;
}

.c2:focus + span {
  color: royalblue;
}

.c2:focus + span::before {
  border-color: royalblue;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="name"
  >
    Label
  </label>
  <label>
    <input
      checked={true}
      className="c2"
      name="name"
      type="checkbox"
      value="one"
    />
    <span
      className="c3"
    >
      Option 1
    </span>
  </label>
  <label>
    <input
      checked={false}
      className="c2"
      name="name"
      type="checkbox"
      value="two"
    />
    <span
      className="c3"
    >
      Option 2
    </span>
  </label>
  <label>
    <input
      checked={true}
      className="c2"
      name="name"
      type="checkbox"
      value="three"
    />
    <span
      className="c3"
    >
      Option 3
    </span>
  </label>
  <span
    className="c4"
  >
    Error message
  </span>
</div>
`;

exports[`<Input /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
}

.c1 {
  font-weight: bold;
}

.c3 {
  color: red;
}

.c2 {
  margin: 0;
  outline: none;
  border: 1px solid #333;
  border-radius: 0;
  background: #fff;
  padding: 4px;
}

.c2:focus {
  border-color: royalblue;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="name"
  >
    Label
  </label>
  <input
    className="c2"
    id="name"
    name="name"
    type="text"
    value="Value"
  />
  <span
    className="c3"
  >
    Error message
  </span>
</div>
`;

exports[`<Radio /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
}

.c1 {
  font-weight: bold;
}

.c4 {
  color: red;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3::before {
  margin: 0;
  outline: none;
  border: 1px solid #333;
  border-radius: 0;
  background: #fff;
  padding: 4px;
  display: block;
  margin-right: 4px;
  padding: 0;
  width: 13px;
  height: 13px;
  content: '';
}

.c3::before:focus {
  border-color: royalblue;
}

.c3::after {
  position: absolute;
  top: calc(50% - 11px / 2);
  left: 2px;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  background: royalblue;
  width: 11px;
  height: 11px;
  content: '';
}

.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2:checked + span::after {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c2 + span::before,
.c2:checked + span::after {
  border-radius: 50%;
}

.c2:focus + span {
  color: royalblue;
}

.c2:focus + span::before {
  border-color: royalblue;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="name"
  >
    Label
  </label>
  <label>
    <input
      checked={false}
      className="c2"
      name="name"
      type="radio"
      value="one"
    />
    <span
      className="c3"
    >
      Option 1
    </span>
  </label>
  <label>
    <input
      checked={true}
      className="c2"
      name="name"
      type="radio"
      value="two"
    />
    <span
      className="c3"
    >
      Option 2
    </span>
  </label>
  <label>
    <input
      checked={false}
      className="c2"
      name="name"
      type="radio"
      value="three"
    />
    <span
      className="c3"
    >
      Option 3
    </span>
  </label>
  <span
    className="c4"
  >
    Error message
  </span>
</div>
`;

exports[`<Select /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
}

.c1 {
  font-weight: bold;
}

.c4 {
  color: red;
}

.c3 {
  margin: 0;
  outline: none;
  border: 1px solid #333;
  border-radius: 0;
  background: #fff;
  padding: 4px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3:focus {
  border-color: royalblue;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c2::after {
  display: block;
  position: absolute;
  top: calc(50% - 5px / 2);
  right: 5px;
  border: 5px solid;
  border-color: #333 transparent transparent transparent;
  pointer-events: none;
  content: '';
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="name"
  >
    Label
  </label>
  <div
    className="c2"
  >
    <select
      className="c3"
      id="name"
      name="name"
      value="two"
    >
      <option
        value="one"
      >
        Option 1
      </option>
      <option
        value="two"
      >
        Option 2
      </option>
      <option
        value="three"
      >
        Option 3
      </option>
    </select>
  </div>
  <span
    className="c4"
  >
    Error message
  </span>
</div>
`;

exports[`<TextArea /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
}

.c1 {
  font-weight: bold;
}

.c3 {
  color: red;
}

.c2 {
  margin: 0;
  outline: none;
  border: 1px solid #333;
  border-radius: 0;
  background: #fff;
  padding: 4px;
}

.c2:focus {
  border-color: royalblue;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="name"
  >
    Label
  </label>
  <textarea
    className="c2"
    id="name"
    name="name"
    value="Value"
  />
  <span
    className="c3"
  >
    Error message
  </span>
</div>
`;
